src/server.rs

```diff
@@ -198,9 +197,19 @@ impl Read for Encoder {
                     // Figure out how many bytes we can read.
                     let upper_bound = (bytes_read + body_len - body_bytes_read).min(buf.len());
                     // Read bytes from body
-                    let new_body_bytes_read =
-                        ready!(Pin::new(&mut self.res)
-                            .poll_read(cx, &mut buf[bytes_read..upper_bound]))?;
+                    let inner_poll_result =
+                        Pin::new(&mut self.res).poll_read(cx, &mut buf[bytes_read..upper_bound]);
+                    let new_body_bytes_read = match inner_poll_result {
+                        Poll::Ready(Ok(n)) => n,
+                        Poll::Pending => {
+                            if bytes_read == 0 {
+                                return Poll::Pending;
+                            } else {
+                                break;
+                            }
+                        }
+                        e => return e,
```

Comment:
Even though this is more typing, this should be easier to follow since all left-hand match arms then use fully qualified types, rather than a named catch-all argument as is the case now:

```suggestion
                        Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
```